src/client/datascience/jupyter/kernels/kernel.ts

```diff
@@ -417,8 +428,16 @@ export class Kernel implements IKernel {
                 // this.kernelExecution.cancel();
                 // Ignore when notebook is disposed as a result of failed restarts.
                 if (!this._ignoreNotebookDisposedErrors) {
+                    const isActiveNotebookDead = this.notebook === notebook;
+
                     this._notebookPromise = undefined;
-                    this._onDisposed.fire();
+                    this.notebook = undefined;
+
+                    // If the active notebook died, then kernel is dead.
+                    if (isActiveNotebookDead) {
+                        this.isKernelDead = true;
```

Comment:
Ensure we don't trigger eevnts unnecessarily (e.g. if somehow create a new INotebook, then we need to take those into account)